### PR TITLE
drivers: hwinfo: sam: Introduce RSTC driver

### PR DESCRIPTION
**Pull request by pdgendt**
- Add a new driver to get the reset cause for some Atmel SAM SoCs. 
- Add a `dts` binding to allow external user reset on the NRST pin.
- Enable driver on Atmel SAM boards

### CODE REVIEW

File: drivers/hwinfo/hwinfo_sam_rstc.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	/* Enable RSTC in PMC */
+	soc_pmc_peripheral_enable(DT_INST_PROP(0, peripheral_id));
+
+	/* Get current Mode Register value */
+	mode = regs->RSTC_MR;
+
+	/* Enable/disable user reset on NRST */
+#if DT_INST_PROP(0, user_nrst)
+	mode &= ~RSTC_MR_KEY_Msk;
+	mode |= (RSTC_MR_URSTEN | RSTC_MR_KEY_PASSWD);
```
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_sam_rstc.c:72]: This is the first driver that actually changes register.

**Review by alexanderwachter — APPROVED** [review]

File: drivers/hwinfo/Kconfig [hunk 2]
```diff
@@ -70,12 +70,20 @@ config HWINFO_IMXRT
 	help
 	  Enable NXP i.mx RT hwinfo driver.
 
+config HWINFO_SAM_RSTC
+	bool "Atmel SAM reset cause"
+	default y
+	depends on SOC_SERIES_SAM4S || SOC_SERIES_SAME70 || \
+		   SOC_SERIES_SAMV71
```
Comment (nandojve) [inline drivers/hwinfo/Kconfig:77]: I think only SoC version that not uses this driver is SAM4L.

```diff
-	depends on SOC_SERIES_SAM4S || SOC_SERIES_SAME70 || \
-		   SOC_SERIES_SAMV71
+	depends on SOC_FAMILY_SAM && !SOC_SERIES_SAM4L
`````

**Review by nandojve — CHANGES_REQUESTED** [review]

File: drivers/hwinfo/Kconfig (hunk 2 above)
Comment (pdgendt) [reply drivers/hwinfo/Kconfig:77]: The other SoCs are missing defines in the RSTC header file, so the driver didn’t compile for those.

Comment (nandojve) [reply drivers/hwinfo/Kconfig:77]: Hi @pdgendt ,
If you don't mind, could you fix it at hal_atmel? This is not a block, it is just to get the complete feature on all supported SoC.

Comment (pdgendt) [reply drivers/hwinfo/Kconfig:77]: I’m currently on vacation without a development setup. It can be a separate PR, or it’ll have to wait. 😅

Comment (nandojve) [reply drivers/hwinfo/Kconfig:77]: No problem, fell free to update in the future. It is a nice to have.

**Review by nandojve — APPROVED** [review]

**Pull request merged**